### PR TITLE
Translation services

### PR DESCRIPTION
**Pull request by vickyokrm**
Includes additional translation service providers to the RC.

### CODE REVIEW

File: .docker/Dockerfile.local [hunk 1]
```diff
@@ -1,20 +1,55 @@
...
+# Add Chatpal which is a submodule
+WORKDIR /app
+RUN meteor npm i \
+    && meteor npm run postinstall \
+    && set +e \
+    && meteor add rocketchat:lib --allow-superuser \
+    && set -e \
+    && meteor build --allow-superuser --server-only --headless --directory /tmp/build
+
+FROM assistify/chat-base:latest
```
Comment (mrsimpson) [inline .docker/Dockerfile.local]: I've seen this image is basically copy-paste from the RC base - except
`ENV DEBIAN_FRONTEND noninteractive` and an `apt-get update`.
Can you explain some motivations about that?
Also note, that using the ENV is discouraged in favor of inline args (https://serverfault.com/a/797318)

File: packages/assistify-deepl-translation/server/deeplTranslate.js [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+ */
+
+
+import {TranslationProviderRegistry, AutoTranslate} from 'meteor/rocketchat:autotranslate';
+import {RocketChat} from 'meteor/rocketchat:lib';
+import {SystemLogger} from 'meteor/rocketchat:logger';
+import _ from 'underscore';
+
+/**
+ * Represents DEEPL translate class
```
Comment (mrsimpson) [inline packages/assistify-deepl-translation/server/deeplTranslate.js]: now *that's* what I call a verbose class documentation.

File: packages/assistify-deepl-translation/server/deeplTranslate.js [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+	 * setup api reference to deepl translate to be used as message translation provider.
+	 * @constructor
+	 */
+	constructor() {
+		super();
+		this.name = 'deepl-translate';
+		this.apiEndPointUrl = 'https://api.deepl.com/v1/translate';
+		// self register & de-register callback - afterSaveMessage based on the activeProvider
+		RocketChat.settings.get('AutoTranslate_ServiceProvider', (key, value) => {
+			if (this.name === value) {
```
Comment (mrsimpson) [inline packages/assistify-deepl-translation/server/deeplTranslate.js]: What for?

File: packages/assistify-dbs-translation/server/dbsTranslate.js [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+import {TranslationProviderRegistry, AutoTranslate} from 'meteor/rocketchat:autotranslate';
+import {RocketChat} from 'meteor/rocketchat:lib';
+import {SystemLogger} from 'meteor/rocketchat:logger';
+import _ from 'underscore';
+
+const cld = Npm.require('cld'); // import the local package dependencies
+
+/**
+ * Intergrate DBS translation service
```
Comment (mrsimpson) [inline packages/assistify-dbs-translation/server/dbsTranslate.js]: Don't be too verbose...

File: packages/assistify-dbs-translation/server/dbsTranslate.js [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+	constructor() {
+		super();
+		this.name = 'dbs-translate';
+		// get the service provider url from settings.
+		RocketChat.settings.get('AutoTranslate_ServiceProviderURL', (key, value) => {
+			this.apiEndPointUrl = value;
+		});
+		// self register & de-register callback - afterSaveMessage based on the activeProvider
+		RocketChat.settings.get('AutoTranslate_ServiceProvider', (key, value) => {
+			if (this.name === value) {
```
Comment (mrsimpson) [inline packages/assistify-dbs-translation/server/dbsTranslate.js]: why?

File: packages/assistify-dbs-translation/server/dbsTranslate.js [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+	 * @param {object} targetLanguages
+	 * @returns {object} translations: Translated messages for each language
+	 */
+	_sendRequestTranslateMessage(message, targetLanguages) {
+		const translations = {};
+		const msgs = message.msg.split('\n');
+		const query = msgs.join();
+		let sourceLanguage;
+		/**
+		 * Service provider do not handle the text language detection automatically rather it requires the source language to be specified
```
Comment (mrsimpson) [inline packages/assistify-dbs-translation/server/dbsTranslate.js]: 👍 excellent comment

File: packages/assistify-dbs-translation/server/dbsTranslate.js [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+						}, data: {
+							text: query,
+							to: language,
+							from: sourceLanguage
+						}
+					});
+					translations[language] = this.deTokenize(Object.assign({}, message, {msg: result}));
+				});
+			} else {
+				SystemLogger.log('Text language could not be determined', err.message);
```
Comment (mrsimpson) [inline packages/assistify-dbs-translation/server/dbsTranslate.js]: this should be a warning

File: packages/assistify-dbs-translation/server/dbsTranslate.js [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+							to: language,
+							from: sourceLanguage
+						}
+					});
+				});
+			} else {
+				SystemLogger.log('Text language could not be determined', err.message);
+			}
+		});
+		return translations;
```
Comment (mrsimpson) [inline packages/assistify-dbs-translation/server/dbsTranslate.js]: this return is outside the callback => should this be `await`ed (return a Promise)?

File: packages/assistify-dbs-translation/server/methods/translate.js [hunk 9]
```diff
@@ -0,0 +1,13 @@
+
+Meteor.methods({
+	'assistify.translate'(httpMethod, url, options) {
```
Comment (mrsimpson) [inline packages/assistify-dbs-translation/server/methods/translate.js]: This is a *very* generic signature which has got nothing to do with autotranslate. Also, it could be used to inject malware from another server (the client could make the server `get` from any address). 
There should be a limitation on the URL.

File: packages/assistify-dbs-translation/server/dbsTranslate.js [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+							from: sourceLanguage
+						}
+					});
+					translations[language] = this.deTokenize(Object.assign({}, message, {msg: result}));
+				});
+			} else {
+				SystemLogger.log('Text language could not be determined', err.message);
+			}
+		});
+		return translations;
```
Comment (mrsimpson) [inline packages/assistify-dbs-translation/server/dbsTranslate.js]: this return is outside the callback => should this be `await`ed (return a Promise)?

File: packages/assistify-deepl-translation/server/deeplTranslate.js [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+			try {
+				const result = HTTP.get(this.apiEndPointUrl, {
+					params: {
+						auth_key: this.apiKey,
+						target_lang: language
+					}, query
+				});
+				if (result.statusCode === 200 && result.data && result.data.translations && Array.isArray(result.data.translations) && result.data.translations.length > 0) {
+					if (result.data.translations.map(translation => translation.detected_source_language).join() !== language) {
+						translations[language] = result.data.translations.map(translation => translation.text).join('\n');
```
Comment (mrsimpson) [inline packages/assistify-deepl-translation/server/deeplTranslate.js]: why join a linebreak?

File: packages/assistify-deepl-translation/server/deeplTranslate.js [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+			}
+		});
+		return translations;
+	}
+}
+
+
+Meteor.startup(() => {
+	TranslationProviderRegistry.registerProvider(new DeeplAutoTranslate());
+	RocketChat.AutoTranslate = TranslationProviderRegistry.getActiveServiceProvider();
```
Comment (mrsimpson) [inline packages/assistify-deepl-translation/server/deeplTranslate.js]: why do you do this in all providers? Should there not be a single hook for that?

File: packages/rocketchat-autotranslate/client/lib/autotranslate.js [hunk 13]
```diff
@@ -41,11 +42,15 @@ RocketChat.AutoTranslate = {
 		Meteor.call('autoTranslate.getSupportedLanguages', 'en', (err, languages) => {
 			this.supportedLanguages = languages || [];
 		});
-
 		Tracker.autorun(() => {
 			if (RocketChat.settings.get('AutoTranslate_Enabled') && RocketChat.authz.hasAtLeastOnePermission(['auto-translate'])) {
 				RocketChat.callbacks.add('renderMessage', (message) => {
-					const subscription = RocketChat.models.Subscriptions.findOne({ rid: message.rid }, { fields: { autoTranslate: 1, autoTranslateLanguage: 1 } });
+					const subscription = RocketChat.models.Subscriptions.findOne({rid: message.rid}, {
```
Comment (mrsimpson) [inline packages/rocketchat-autotranslate/client/lib/autotranslate.js]: Not chaning this reduces the diff and makes it easier to merge. Though it looks less beautiful, kindly minimize the footprint

File: packages/rocketchat-autotranslate/client/lib/autotranslate.js [hunk 14]
```diff
@@ -86,11 +99,16 @@ RocketChat.AutoTranslate = {
 				RocketChat.callbacks.remove('streamMessage', 'autotranslate-stream');
 			}
 		});
+
+		Tracker.autorun(() => {
+			if (RocketChat.settings.get('AutoTranslate_ServiceProvider') && RocketChat.authz.hasAtLeastOnePermission(['auto-translate'])) {
```
Comment (mrsimpson) [inline packages/rocketchat-autotranslate/client/lib/autotranslate.js]: What's the motivation for that?

File: packages/rocketchat-autotranslate/client/lib/autotranslate.js [hunk 15]
```diff
@@ -86,11 +99,16 @@ RocketChat.AutoTranslate = {
...
+		});
 	}
+
 };
 
 Meteor.startup(function() {
-	RocketChat.CachedCollectionManager.onLogin(() => {
-		RocketChat.AutoTranslate.init();
-	});
+	RocketChat.AutoTranslate.init();
```
Comment (mrsimpson) [inline packages/rocketchat-autotranslate/client/lib/autotranslate.js]: Why did you remove the `onLogin` wrapper?

File: packages/rocketchat-autotranslate/server/autotranslate.js [hunk 16]
```diff
@@ -220,38 +206,92 @@ class AutoTranslate {
...
+	 * @private
+	 * @returns {object}
+	 */
+	_getProviderMetadata() {
+		SystemLogger.warn('must be implemented by subclass!', '_getProviderMetadata');
+	}
+
+
+	/**
+	 * Returns necessary settings information about the translation service provider.
```
Comment (mrsimpson) [inline packages/rocketchat-autotranslate/server/autotranslate.js]: Does this match the method's semantics?

File: packages/rocketchat-autotranslate/server/autotranslate.js [hunk 17]
```diff
@@ -220,38 +206,92 @@ class AutoTranslate {
...
+	 * @private
+	 * @param {string} target
+	 * @returns {object}
+	 */
+	_getSupportedLanguages(target) {
+		SystemLogger.warn('must be implemented by subclass!', '_getSupportedLanguages', target);
+	}
+
+	/**
+	 * Send Request REST API call to the service provider.
```
Comment (mrsimpson) [inline packages/rocketchat-autotranslate/server/autotranslate.js]: Does this match the method's semantics?

File: packages/rocketchat-autotranslate/server/autotranslate.js [hunk 18]
```diff
@@ -220,38 +206,92 @@ class AutoTranslate {
...
+	 * @param {object} message
+	 * @param {object} targetLanguages
+	 * @return {object}
+	 */
+	_sendRequestTranslateMessage(message, targetLanguages) {
+		SystemLogger.warn('must be implemented by subclass!', '_sendRequestTranslateMessage', message, targetLanguages);
+	}
+
+	/**
+	 * Send Request REST API call to the service provider.
```
Comment (mrsimpson) [inline packages/rocketchat-autotranslate/server/autotranslate.js]: Particularly for the abstract methods, the description is important. Shall this translate the attachments or only the attachment's description?

File: packages/rocketchat-autotranslate/server/methods/getSupportedLanguages.js [hunk 19]
```diff
@@ -4,7 +4,7 @@ Meteor.methods({
 			throw new Meteor.Error('error-action-not-allowed', 'Auto-Translate is not allowed', { method: 'autoTranslate.saveSettings'});
 		}
 
-		return RocketChat.AutoTranslate.getSupportedLanguages(targetLanguage);
+		return RocketChat.AutoTranslate._getSupportedLanguages(targetLanguage);
```
Comment (mrsimpson) [inline packages/rocketchat-autotranslate/server/methods/getSupportedLanguages.js]: Public methods should not be underscore-prefixed

File: packages/rocketchat-autotranslate/server/methods/refreshProviderSettings.js [hunk 20]
```diff
@@ -0,0 +1,7 @@
+import { TranslationProviderRegistry } from '../autotranslate.js';
+Meteor.methods({
```
Comment (mrsimpson) [inline packages/rocketchat-autotranslate/server/methods/refreshProviderSettings.js]: Why does this need to be exposed to the client?

File: packages/rocketchat-autotranslate/server/settings.js [hunk 21]
```diff
@@ -1,4 +1,49 @@
...
+		group: 'Message',
+		section: 'AutoTranslate',
+		public: true
+	});
+	RocketChat.settings.add('AutoTranslate_ServiceProvider', 'google-translate', {
+		type: 'select',
+		group: 'Message',
+		section: 'AutoTranslate',
+		values: [{
+			key: 'google-translate',
```
Comment (mrsimpson) [inline packages/rocketchat-autotranslate/server/settings.js]: Since they are all referenced from file inside package `rocketchat-translation`, I'd favor to also add the deepl- and dbs-translation provider into this package. Alternatively, the settings had to be created based upon information received from the registry.

File: .docker/Dockerfile.local (hunk 1 above)
Comment (vickyokrm) [reply .docker/Dockerfile.local]: is used as non interactive to suppress the prompt while installing package using `apt-get` command. However, this was resolved using `-y` inline `apt-get` command. So the `ENV DEBIAN_FRONTEND` is not required.

File: packages/assistify-deepl-translation/server/deeplTranslate.js (hunk 3 above)
Comment (vickyokrm) [reply packages/assistify-deepl-translation/server/deeplTranslate.js]: This will to set the active provide context when the provider settings changed.

**Comment by vickyokrm** [comment]
cld.detect is synchronous. But the http call to the translate request can promise?
